#1528 - Registration of multiple ManagedProvider services

adimova opened the pull request:
Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

maggu2810 commented:
I didn't follow the whole discussion, but I am confused by the topic "multiple ManagedProvider services" and <code>cardinality="0..1"</code>.

I am informed about that cardinalities:
https://osgi.org/javadoc/r4v43/cmpn/org/osgi/service/component/annotations/ReferenceCardinality.html

Do you want to support an optional ManagedProvider or is my reading / understanding of this PR completely wrong?

adimova commented:
We want to protect the AbstractRegistry from the possibility that multiple ManagedProvider services can be registered. In the current registry component descriptions, you have a dependency like this:
`<reference bind="addProvider" cardinality="0..n" interface="org.eclipse.smarthome.core.thing.link.ItemChannelLinkProvider"  ..`
which means that you expect to have multiple ItemChannelLinkProvider services, and they are all added in the registry with the addProvider method. The code of the addProvider method assumes that only one of them is an instance of the ManagedProvider interface - but that is not (and can not) be guaranteed. So I added a new dependency:
`<reference bind="setManagedProvider" cardinality="0..1" interface="org.eclipse.smarthome.core.thing.link.ManagedItemChannelLinkProvider" .. `
to explicitly address the ManagedProvider, and removed the instanceof check in addProvider. Now even if there are multiple ManagedItemChannelLinkProvider services, the DS runtime will take care to set only the first one in the registry.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java:
```diff
@@ -191,9 +190,12 @@ protected void addProvider(Provider<E> provider) {
+    }
+
+    protected void setManagedProvider(ManagedProvider<E, K> provider) {
+        if (managedProvider == null) {
```
This check can be removed if we rely on a bug-free DS implementation. Correct?

maggu2810 commented:
@adimova thank you for your explanation. Now it is clear to me.

As far as I can see, all is fine.
I assume you can remove the null check.

Will merge it later.

adimova commented:
The null check is removed

maggu2810 commented:
@adimova I assume you have tested your changes and the build succeeded on your system :wink: 
Can you have a look at:
https://hudson.eclipse.org/smarthome/job/SmartHomeDistribution-Nightly/1072/consoleText

maggu2810 commented:
@adimova I assume you have tested your changes and the build succeeded on your system :wink: 
Can you have a look at:
https://hudson.eclipse.org/smarthome/job/SmartHomeDistribution-Nightly/1072/consoleText

Do you think that could be related to your changes?

adimova commented:
Yes Markus, i have tested my changes and the build succeeded. I will check the problem

maggu2810 commented:
@adimova  The build today works, so perhaps it is not really related to your change. Perhaps the test is a little bit unstable, timing dependent or something else.
